In [76]:
import pandas as pd
from pathlib import Path

NOTEBOOK_DIR = Path(".").resolve()
SAMPLE_1_DIR = NOTEBOOK_DIR.parent / "data" / "Sample 1"
SAMPLE_2_DIR = NOTEBOOK_DIR.parent / "data" / "Sample 2"
DATA_DIR = SAMPLE_1_DIR

In [77]:
sample_1 = pd.read_csv('DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv')
sample_2 = pd.read_csv('DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv')

/var/folders/f5/dy840xrs0dq70k185zls6qmc0000gn/T/ipykernel_6481/1591897906.py:2: DtypeWarning: Columns (0: ICD9_DGNS_CD_6, 1: ICD9_DGNS_CD_7, 2: ICD9_DGNS_CD_8, 3: HCPCS_CD_9, 4: HCPCS_CD_10, 5: HCPCS_CD_11, 6: HCPCS_CD_12, 7: HCPCS_CD_13, 8: LINE_ICD9_DGNS_CD_7, 9: LINE_ICD9_DGNS_CD_8, 10: LINE_ICD9_DGNS_CD_9, 11: LINE_ICD9_DGNS_CD_10, 12: LINE_ICD9_DGNS_CD_11, 13: LINE_ICD9_DGNS_CD_12, 14: LINE_ICD9_DGNS_CD_13) have mixed types. Specify dtype option on import or set low_memory=False.
  sample_2 = pd.read_csv('DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv')


In [78]:
sample_1.shape


(116352, 32)

In [79]:
sample_1.columns.tolist()
sample_2.columns.tolist()

['DESYNPUF_ID',
 'CLM_ID',
 'CLM_FROM_DT',
 'CLM_THRU_DT',
 'ICD9_DGNS_CD_1',
 'ICD9_DGNS_CD_2',
 'ICD9_DGNS_CD_3',
 'ICD9_DGNS_CD_4',
 'ICD9_DGNS_CD_5',
 'ICD9_DGNS_CD_6',
 'ICD9_DGNS_CD_7',
 'ICD9_DGNS_CD_8',
 'PRF_PHYSN_NPI_1',
 'PRF_PHYSN_NPI_2',
 'PRF_PHYSN_NPI_3',
 'PRF_PHYSN_NPI_4',
 'PRF_PHYSN_NPI_5',
 'PRF_PHYSN_NPI_6',
 'PRF_PHYSN_NPI_7',
 'PRF_PHYSN_NPI_8',
 'PRF_PHYSN_NPI_9',
 'PRF_PHYSN_NPI_10',
 'PRF_PHYSN_NPI_11',
 'PRF_PHYSN_NPI_12',
 'PRF_PHYSN_NPI_13',
 'TAX_NUM_1',
 'TAX_NUM_2',
 'TAX_NUM_3',
 'TAX_NUM_4',
 'TAX_NUM_5',
 'TAX_NUM_6',
 'TAX_NUM_7',
 'TAX_NUM_8',
 'TAX_NUM_9',
 'TAX_NUM_10',
 'TAX_NUM_11',
 'TAX_NUM_12',
 'TAX_NUM_13',
 'HCPCS_CD_1',
 'HCPCS_CD_2',
 'HCPCS_CD_3',
 'HCPCS_CD_4',
 'HCPCS_CD_5',
 'HCPCS_CD_6',
 'HCPCS_CD_7',
 'HCPCS_CD_8',
 'HCPCS_CD_9',
 'HCPCS_CD_10',
 'HCPCS_CD_11',
 'HCPCS_CD_12',
 'HCPCS_CD_13',
 'LINE_NCH_PMT_AMT_1',
 'LINE_NCH_PMT_AMT_2',
 'LINE_NCH_PMT_AMT_3',
 'LINE_NCH_PMT_AMT_4',
 'LINE_NCH_PMT_AMT_5',
 'LINE_NCH_PMT_AMT_6',
 '

In [80]:
sample_1.head()


,DESYNPUF_ID,BENE_BIRTH_DT,BENE_DEATH_DT,BENE_SEX_IDENT_CD,BENE_RACE_CD,BENE_ESRD_IND,SP_STATE_CODE,BENE_COUNTY_CD,BENE_HI_CVRAGE_TOT_MONS,BENE_SMI_CVRAGE_TOT_MONS,...,SP_STRKETIA,MEDREIMB_IP,BENRES_IP,PPPYMT_IP,MEDREIMB_OP,BENRES_OP,PPPYMT_OP,MEDREIMB_CAR,BENRES_CAR,PPPYMT_CAR
0,00013D2EFD8E45D1,19230501,NaN,1,1,0,26,950,12,12,...,2,0.0,0.0,0.0,50.0,10.0,0.0,0.0,0.0,0.0
1,00016F745862898F,19430101,NaN,1,1,0,39,230,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,700.0,240.0,0.0
2,0001FDD721E223DC,19360901,NaN,2,1,0,39,280,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,00021CA6FF03E670,19410601,NaN,1,5,0,6,290,0,0,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,00024B3D2352D2D0,19360801,NaN,1,1,0,52,590,12,12,...,2,0.0,0.0,0.0,30.0,40.0,0.0,220.0,80.0,0.0


In [81]:
sample_1.info()

<class 'pandas.DataFrame'>
RangeIndex: 116352 entries, 0 to 116351
Data columns (total 32 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   DESYNPUF_ID               116352 non-null  str    
 1   BENE_BIRTH_DT             116352 non-null  int64  
 2   BENE_DEATH_DT             1814 non-null    float64
 3   BENE_SEX_IDENT_CD         116352 non-null  int64  
 4   BENE_RACE_CD              116352 non-null  int64  
 5   BENE_ESRD_IND             116352 non-null  str    
 6   SP_STATE_CODE             116352 non-null  int64  
 7   BENE_COUNTY_CD            116352 non-null  int64  
 8   BENE_HI_CVRAGE_TOT_MONS   116352 non-null  int64  
 9   BENE_SMI_CVRAGE_TOT_MONS  116352 non-null  int64  
 10  BENE_HMO_CVRAGE_TOT_MONS  116352 non-null  int64  
 11  PLAN_CVRG_MOS_NUM         116352 non-null  int64  
 12  SP_ALZHDMTA               116352 non-null  int64  
 13  SP_CHF                    116352 non-null  int64  
 14 

In [82]:
sample_1.describe()

,BENE_BIRTH_DT,BENE_DEATH_DT,BENE_SEX_IDENT_CD,BENE_RACE_CD,SP_STATE_CODE,BENE_COUNTY_CD,BENE_HI_CVRAGE_TOT_MONS,BENE_SMI_CVRAGE_TOT_MONS,BENE_HMO_CVRAGE_TOT_MONS,PLAN_CVRG_MOS_NUM,...,SP_STRKETIA,MEDREIMB_IP,BENRES_IP,PPPYMT_IP,MEDREIMB_OP,BENRES_OP,PPPYMT_OP,MEDREIMB_CAR,BENRES_CAR,PPPYMT_CAR
count,1.163520e+05,1.814000e+03,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,...,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000,116352.000000
mean,1.936418e+07,2.008065e+07,1.553037,1.284662,25.693740,366.486188,11.143994,10.495514,2.576183,6.826466,...,1.955110,2214.180762,249.053441,99.142258,622.226520,197.502235,25.724182,1162.095881,328.747508,18.355851
std,1.250446e+05,3.493908e+02,0.497181,0.754885,15.582601,266.106008,2.839995,3.758701,4.828831,5.814787,...,0.207063,8473.340573,885.356400,1857.930522,1796.476653,522.437900,370.974324,1587.643182,436.858441,87.356293
min,1.909010e+07,2.008010e+07,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,1.000000,-3000.000000,0.000000,0.000000,-90.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.928110e+07,2.008030e+07,1.000000,1.000000,10.000000,140.000000,12.000000,12.000000,0.000000,0.000000,...,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,1.936050e+07,2.008070e+07,2.000000,1.000000,25.000000,330.000000,12.000000,12.000000,0.000000,12.000000,...,2.000000,0.000000,0.000000,0.000000,20.000000,0.000000,0.000000,610.000000,170.000000,0.000000
75%,1.942030e+07,2.008100e+07,2.000000,1.000000,39.000000,560.000000,12.000000,12.000000,0.000000,12.000000,...,2.000000,0.000000,0.000000,0.000000,550.000000,180.000000,0.000000,1650.000000,480.000000,0.000000
max,1.983120e+07,2.008120e+07,2.000000,5.000000,54.000000,999.000000,12.000000,12.000000,12.000000,12.000000,...,2.000000,164220.000000,53096.000000,68000.000000,50020.000000,12450.000000,14400.000000,21160.000000,5260.000000,2110.000000


## DESYNPUF_ID uniqueness

Check whether beneficiary IDs are unique within each file and whether Sample 1 and Sample 2 share the same IDs.

In [83]:
def check_id_uniqueness(csv_path):
    """Return row count, unique ID count, and whether DESYNPUF_ID is unique per row."""
    ids = pd.read_csv(csv_path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID']
    n_rows = len(ids)
    n_unique = ids.nunique()
    return {
        'file': csv_path.name,
        'rows': n_rows,
        'unique_ids': n_unique,
        'duplicates': n_rows - n_unique,
        'is_unique': n_rows == n_unique,
    }

In [84]:
# Beneficiary summary files: one row per beneficiary
sample_1_bene = pd.read_csv(
    'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv',
    usecols=['DESYNPUF_ID'],
    dtype=str,
)

n_rows = len(sample_1_bene)
n_unique = sample_1_bene['DESYNPUF_ID'].nunique()

print(f'Sample 1 (2008 Beneficiary Summary)')
print(f'  Rows: {n_rows:,}')
print(f'  Unique DESYNPUF_ID: {n_unique:,}')
print(f'  All IDs unique: {n_rows == n_unique}')

Sample 1 (2008 Beneficiary Summary)
  Rows: 116,352
  Unique DESYNPUF_ID: 116,352
  All IDs unique: True


In [85]:
# All Sample 1 files
sample_1_results = pd.DataFrame(
    check_id_uniqueness(f) for f in sorted(SAMPLE_1_DIR.glob('*.csv'))
)
sample_1_results

,file,rows,unique_ids,duplicates,is_unique
0,DE1_0_2008_Beneficiary_Summary_File_Sample_1 2...,116352,116352,0,True
1,DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv,2370667,49278,2321389,False
2,DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv,2370668,49349,2321319,False
3,DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv,66773,37780,28993,False
4,DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv,790790,85272,705518,False
5,DE1_0_2008_to_2010_Prescription_Drug_Events_Sa...,5552421,99538,5452883,False
6,DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv,114538,114538,0,True
7,DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv,112754,112754,0,True


In [86]:
# All Sample 2 files
sample_2_results = pd.DataFrame(
    check_id_uniqueness(f) for f in sorted(SAMPLE_2_DIR.glob('*.csv'))
)
sample_2_results

,file,rows,unique_ids,duplicates,is_unique
0,DE1_0_2008_Beneficiary_Summary_File_Sample_2.csv,116395,116395,0,True
1,DE1_0_2008_to_2010_Carrier_Claims_Sample_2A.csv,2372957,49153,2323804,False
2,DE1_0_2008_to_2010_Carrier_Claims_Sample_2B.csv,2372957,49534,2323423,False
3,DE1_0_2008_to_2010_Inpatient_Claims_Sample_2.csv,66494,37661,28833,False
4,DE1_0_2008_to_2010_Outpatient_Claims_Sample_2.csv,792562,85295,707267,False
5,DE1_0_2008_to_2010_Prescription_Drug_Events_Sa...,5561154,99746,5461408,False
6,DE1_0_2009_Beneficiary_Summary_File_Sample_2.csv,114618,114618,0,True
7,DE1_0_2010_Beneficiary_Summary_File_Sample_2.csv,112845,112845,0,True


In [87]:
def collect_ids(folder):
    """Collect all unique DESYNPUF_ID values from every CSV in a sample folder."""
    all_ids = set()
    for csv_path in sorted(folder.glob('*.csv')):
        ids = pd.read_csv(csv_path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID']
        all_ids.update(ids)
    return all_ids

sample_1_ids = collect_ids(SAMPLE_1_DIR)
sample_2_ids = collect_ids(SAMPLE_2_DIR)

overlap = sample_1_ids & sample_2_ids

print(f'Unique IDs in Sample 1: {len(sample_1_ids):,}')
print(f'Unique IDs in Sample 2: {len(sample_2_ids):,}')
print(f'Overlapping IDs: {len(overlap):,}')
print(f'Same ID sets: {sample_1_ids == sample_2_ids}')

Unique IDs in Sample 1: 116,352
Unique IDs in Sample 2: 116,395
Overlapping IDs: 0
Same ID sets: False


### Summary

- **Beneficiary Summary files**: `DESYNPUF_ID` is unique (one row per person).
- **Claims / prescription files**: IDs repeat because each beneficiary has many claims/events.
- **Sample 1 vs Sample 2**: The two samples contain different beneficiaries with no overlapping IDs.

## Can `DESYNPUF_ID` from the Beneficiary Summary be used as a primary key?

For Sample 1, check whether the 2008 Beneficiary Summary IDs are:
1. **Unique** in the beneficiary file (valid primary key on the parent table)
2. **Referenced correctly** in all other Sample 1 files (every ID in claims/PDE files exists in the beneficiary summary)

In [88]:
# Primary key set from the 2008 Beneficiary Summary (sample_1)
bene_pk = set(sample_1['DESYNPUF_ID'])

print(f'Beneficiary summary rows: {len(sample_1):,}')
print(f'Unique DESYNPUF_ID: {sample_1["DESYNPUF_ID"].nunique():,}')
print(f'Valid primary key (no duplicates): {len(sample_1) == sample_1["DESYNPUF_ID"].nunique()}')

Beneficiary summary rows: 116,352
Unique DESYNPUF_ID: 116,352
Valid primary key (no duplicates): True


In [89]:
BENE_FILE = 'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv'

other_files = sorted(
    f for f in SAMPLE_1_DIR.glob('*.csv')
    if f.name != BENE_FILE
)

[f.name for f in other_files]

['DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv',
 'DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv',
 'DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv',
 'DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv',
 'DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv',
 'DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv',
 'DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv']

In [90]:
pk_checks = []

for csv_path in other_files:
    file_ids = set(
        pd.read_csv(csv_path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID']
    )
    orphans = file_ids - bene_pk
    matched = file_ids & bene_pk

    pk_checks.append({
        'file': csv_path.name,
        'unique_ids_in_file': len(file_ids),
        'ids_found_in_bene_summary': len(matched),
        'orphan_ids_not_in_bene': len(orphans),
        'all_ids_in_bene_summary': len(orphans) == 0,
        'bene_ids_missing_from_file': len(bene_pk - file_ids),
        'pct_bene_with_records_in_file': round(100 * len(matched) / len(bene_pk), 2),
    })

pk_check_df = pd.DataFrame(pk_checks)
pk_check_df

,file,unique_ids_in_file,ids_found_in_bene_summary,orphan_ids_not_in_bene,all_ids_in_bene_summary,bene_ids_missing_from_file,pct_bene_with_records_in_file
0,DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv,49278,49278,0,True,67074,42.35
1,DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv,49349,49349,0,True,67003,42.41
2,DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv,37780,37780,0,True,78572,32.47
3,DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv,85272,85272,0,True,31080,73.29
4,DE1_0_2008_to_2010_Prescription_Drug_Events_Sa...,99538,99538,0,True,16814,85.55
5,DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv,114538,114538,0,True,1814,98.44
6,DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv,112754,112754,0,True,3598,96.91


In [91]:
all_files_valid = pk_check_df['all_ids_in_bene_summary'].all()

print(f'All other Sample 1 files reference only beneficiary summary IDs: {all_files_valid}')
print(f'Total orphan IDs across all files: {pk_check_df["orphan_ids_not_in_bene"].sum()}')

All other Sample 1 files reference only beneficiary summary IDs: True
Total orphan IDs across all files: 0


### Conclusion

- **`DESYNPUF_ID` is a valid primary key** in the 2008 Beneficiary Summary: one unique ID per row.
- **All other Sample 1 files can be joined to it** on `DESYNPUF_ID`: every ID in the claims and prescription files exists in the beneficiary summary (no orphan IDs).
- **Claims and PDE files are not one row per ID** — the same beneficiary appears on many rows. Use `DESYNPUF_ID` as a **foreign key** in those files, not as their primary key.
- **Not every beneficiary appears in every file** — some people have no carrier, inpatient, outpatient, or drug events. That is expected; it does not break the join.

## Do IDs match across files and sample folders?

Check whether `DESYNPUF_ID` values are the **same set** in every file within Sample 1, and whether different sample folders (Sample 1, Sample 2, …) share any IDs.

In [92]:
def load_ids(csv_path):
  return set(pd.read_csv(csv_path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])

DATA_ROOT = SAMPLE_1_DIR.parent
sample_1_files = sorted(SAMPLE_1_DIR.glob('*.csv'))

file_id_sets = {f.name: load_ids(f) for f in sample_1_files}
id_counts = pd.Series({name: len(ids) for name, ids in file_id_sets.items()}, name='unique_ids')
id_counts

DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv          116352
DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv              49278
DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv              49349
DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv             37780
DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv            85272
DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv     99538
DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv            114538
DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv            112754
Name: unique_ids, dtype: int64

In [93]:
# Are the ID sets identical across every Sample 1 file?
unique_sets = {frozenset(ids) for ids in file_id_sets.values()}

print(f'Number of distinct ID sets in Sample 1: {len(unique_sets)}')
print(f'All Sample 1 files have the exact same IDs: {len(unique_sets) == 1}')
print(f'Union of IDs across all Sample 1 files: {len(set().union(*file_id_sets.values())):,}')

Number of distinct ID sets in Sample 1: 8
All Sample 1 files have the exact same IDs: False
Union of IDs across all Sample 1 files: 116,352


In [94]:
BENE_2008_FILE = 'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv'
bene_2008_ids = file_id_sets[BENE_2008_FILE]

cross_file_checks = []

for file_name, file_ids in file_id_sets.items():
    if file_name == BENE_2008_FILE:
        continue

    cross_file_checks.append({
        'file': file_name,
        'unique_ids': len(file_ids),
        'ids_in_2008_bene': len(file_ids & bene_2008_ids),
        'ids_not_in_2008_bene': len(file_ids - bene_2008_ids),
        'bene_ids_missing_from_file': len(bene_2008_ids - file_ids),
        'exact_same_set_as_2008_bene': file_ids == bene_2008_ids,
        'all_file_ids_in_2008_bene': file_ids <= bene_2008_ids,
    })

cross_file_df = pd.DataFrame(cross_file_checks)
cross_file_df

,file,unique_ids,ids_in_2008_bene,ids_not_in_2008_bene,bene_ids_missing_from_file,exact_same_set_as_2008_bene,all_file_ids_in_2008_bene
0,DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv,49278,49278,0,67074,False,True
1,DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv,49349,49349,0,67003,False,True
2,DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv,37780,37780,0,78572,False,True
3,DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv,85272,85272,0,31080,False,True
4,DE1_0_2008_to_2010_Prescription_Drug_Events_Sa...,99538,99538,0,16814,False,True
5,DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv,114538,114538,0,1814,False,True
6,DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv,112754,112754,0,3598,False,True


In [95]:
print('Every non-beneficiary file is a subset of the 2008 beneficiary IDs:',
      cross_file_df['all_file_ids_in_2008_bene'].all())
print('Any file has IDs outside the 2008 beneficiary file:',
      (cross_file_df['ids_not_in_2008_bene'] > 0).any())
print('Any file has the exact same ID set as the 2008 beneficiary file:',
      cross_file_df['exact_same_set_as_2008_bene'].any())

Every non-beneficiary file is a subset of the 2008 beneficiary IDs: True
Any file has IDs outside the 2008 beneficiary file: False
Any file has the exact same ID set as the 2008 beneficiary file: False


### IDs across sample folders (Sample 1, Sample 2, …)

Each sample folder is a separate synthetic cohort. Compare 2008 Beneficiary Summary files across all sample folders.

In [96]:
sample_folder_ids = {}

for folder in sorted(DATA_ROOT.glob('Sample *'), key=lambda p: int(p.name.split()[-1])):
    bene_file = next(folder.glob('*2008*Beneficiary*'))
    sample_folder_ids[folder.name] = load_ids(bene_file)

sample_folder_counts = pd.Series(
    {name: len(ids) for name, ids in sample_folder_ids.items()},
    name='unique_ids_2008_bene',
)
sample_folder_counts

Sample 1     116352
Sample 2     116395
Sample 3     116390
Sample 4     116279
Sample 5     116364
Sample 6     116234
Sample 7     116352
Sample 8     116330
Sample 9     116287
Sample 10    116353
Sample 11    116339
Sample 12    116387
Sample 13    116363
Sample 14    116374
Sample 15    116391
Sample 16    116377
Sample 17    116285
Sample 18    116405
Sample 19    116224
Sample 20    116375
Name: unique_ids_2008_bene, dtype: int64

In [97]:
folder_names = list(sample_folder_ids.keys())
overlap_rows = []

for i, folder_a in enumerate(folder_names):
    for folder_b in folder_names[i + 1:]:
        overlap = sample_folder_ids[folder_a] & sample_folder_ids[folder_b]
        overlap_rows.append({
            'folder_a': folder_a,
            'folder_b': folder_b,
            'shared_ids': len(overlap),
        })

overlap_df = pd.DataFrame(overlap_rows)
overlap_df[overlap_df['shared_ids'] > 0]

,folder_a,folder_b,shared_ids


In [98]:
total_overlaps = overlap_df['shared_ids'].sum()

print(f'Sample folders compared: {len(folder_names)}')
print(f'Pairs with shared IDs: {(overlap_df["shared_ids"] > 0).sum()}')
print(f'Total shared IDs across all folder pairs: {total_overlaps}')
print(f'Sample folders use completely separate ID pools: {total_overlaps == 0}')

Sample folders compared: 20
Pairs with shared IDs: 0
Total shared IDs across all folder pairs: 0
Sample folders use completely separate ID pools: True


### Summary

**Within Sample 1**
- Files do **not** all contain the same ID set — claims and PDE files have fewer unique IDs because not every beneficiary has every type of claim.
- Every ID in every Sample 1 file **does appear** in the 2008 Beneficiary Summary (no mismatched/orphan IDs).
- The 2008 Beneficiary Summary is the **master list** of 116,352 beneficiaries for Sample 1.

**Across sample folders**
- Sample 1, Sample 2, … Sample 20 are **different cohorts** with **no shared `DESYNPUF_ID` values**.
- IDs are safe to use for joins **within** a sample folder, but not **across** sample folders.

## Primary key in the 2008 Beneficiary Summary

A primary key must uniquely identify each row (no duplicates, no nulls).

In [99]:
bene_2008 = sample_1  # 2008 Beneficiary Summary

n_rows = len(bene_2008)
n_unique_desynpuf = bene_2008['DESYNPUF_ID'].nunique()
n_null_desynpuf = bene_2008['DESYNPUF_ID'].isna().sum()
n_duplicate_desynpuf = n_rows - n_unique_desynpuf

print('Candidate primary key: DESYNPUF_ID')
print(f'  Total rows:           {n_rows:,}')
print(f'  Unique values:        {n_unique_desynpuf:,}')
print(f'  Null values:          {n_null_desynpuf:,}')
print(f'  Duplicate values:     {n_duplicate_desynpuf:,}')
print(f'  Is valid primary key: {n_duplicate_desynpuf == 0 and n_null_desynpuf == 0}')

Candidate primary key: DESYNPUF_ID
  Total rows:           116,352
  Unique values:        116,352
  Null values:          0
  Duplicate values:     0
  Is valid primary key: True


In [100]:
# Check whether any other single column is also unique
pk_candidates = []

for col in bene_2008.columns:
    pk_candidates.append({
        'column': col,
        'unique_values': bene_2008[col].nunique(dropna=False),
        'null_values': bene_2008[col].isna().sum(),
        'is_unique': bene_2008[col].nunique(dropna=False) == n_rows,
    })

pk_candidate_df = pd.DataFrame(pk_candidates).sort_values(
    ['is_unique', 'unique_values'], ascending=[False, False]
)
pk_candidate_df

,column,unique_values,null_values,is_unique
0,DESYNPUF_ID,116352,0,True
23,MEDREIMB_IP,2025,0,False
26,MEDREIMB_OP,1466,0,False
29,MEDREIMB_CAR,1222,0,False
1,BENE_BIRTH_DT,900,0,False
27,BENRES_OP,656,0,False
30,BENRES_CAR,399,0,False
7,BENE_COUNTY_CD,307,0,False
24,BENRES_IP,145,0,False
31,PPPYMT_CAR,138,0,False


In [101]:
if n_duplicate_desynpuf > 0:
    bene_2008['DESYNPUF_ID'].value_counts().head(10)
else:
    print('No duplicate DESYNPUF_ID values found.')

No duplicate DESYNPUF_ID values found.


### Answer

The primary key of the 2008 Beneficiary Summary is **`DESYNPUF_ID`**.

- One row per beneficiary
- 116,352 rows and 116,352 unique, non-null `DESYNPUF_ID` values
- This is the CMS synthetic beneficiary identifier and should be used to join to claims and prescription files in the same sample folder

## How are the Sample 1 files different from each other?

Sample 1 contains **8 CSV files** in three groups: beneficiary summaries (demographics & annual totals), claims (medical events), and prescription drug events. They share `DESYNPUF_ID` but differ in **row grain**, **columns**, **time coverage**, and **purpose**.

In [102]:
FILE_GROUPS = {
    'Beneficiary Summary': [
        'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv',
        'DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv',
        'DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv',
    ],
    'Carrier Claims (Part B)': [
        'DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv',
        'DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv',
    ],
    'Institutional Claims': [
        'DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv',
        'DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv',
    ],
    'Prescription Drug Events (Part D)': [
        'DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv',
    ],
}

for group, files in FILE_GROUPS.items():
    print(f'{group}:')
    for f in files:
        print(f'  - {f}')

Beneficiary Summary:
  - DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv
  - DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv
  - DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv
Carrier Claims (Part B):
  - DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv
  - DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv
Institutional Claims:
  - DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv
  - DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv
Prescription Drug Events (Part D):
  - DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv


In [103]:
def file_overview(csv_path):
    header = pd.read_csv(csv_path, nrows=0)
    ids = pd.read_csv(csv_path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID']

    row = {
        'file': csv_path.name,
        'rows': len(ids),
        'columns': len(header.columns),
        'unique_DESYNPUF_ID': ids.nunique(),
        'rows_per_beneficiary': round(len(ids) / ids.nunique(), 2) if ids.nunique() else None,
    }

    if 'CLM_ID' in header.columns:
        clm = pd.read_csv(csv_path, usecols=['CLM_ID'], dtype=str)['CLM_ID']
        row['grain'] = 'claim (CLM_ID)'
        row['unique_CLM_ID'] = clm.nunique()
    elif 'PDE_ID' in header.columns:
        pde = pd.read_csv(csv_path, usecols=['PDE_ID'], dtype=str)['PDE_ID']
        row['grain'] = 'prescription fill (PDE_ID)'
        row['unique_PDE_ID'] = pde.nunique()
    else:
        row['grain'] = 'beneficiary (DESYNPUF_ID)'
        row['unique_CLM_ID'] = None
        row['unique_PDE_ID'] = None

    return row

sample_1_overview = pd.DataFrame(
    file_overview(SAMPLE_1_DIR / f) for f in sorted(SAMPLE_1_DIR.glob('*.csv'))
)
sample_1_overview

,file,rows,columns,unique_DESYNPUF_ID,rows_per_beneficiary,grain,unique_CLM_ID,unique_PDE_ID
0,DE1_0_2008_Beneficiary_Summary_File_Sample_1 2...,116352,32,116352,1.00,beneficiary (DESYNPUF_ID),NaN,NaN
1,DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv,2370667,142,49278,48.11,claim (CLM_ID),2370667.0,NaN
2,DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv,2370668,142,49349,48.04,claim (CLM_ID),2370668.0,NaN
3,DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv,66773,81,37780,1.77,claim (CLM_ID),66705.0,NaN
4,DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv,790790,76,85272,9.27,claim (CLM_ID),779815.0,NaN
5,DE1_0_2008_to_2010_Prescription_Drug_Events_Sa...,5552421,8,99538,55.78,prescription fill (PDE_ID),NaN,5552421.0
6,DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv,114538,32,114538,1.00,beneficiary (DESYNPUF_ID),NaN,NaN
7,DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv,112754,32,112754,1.00,beneficiary (DESYNPUF_ID),NaN,NaN


### 1. Beneficiary Summary files (2008, 2009, 2010)

Same 32 columns each year. One row per beneficiary per year. Values change year to year (coverage months, chronic conditions, payments).

In [104]:
bene_files = {
    '2008': SAMPLE_1_DIR / 'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv',
    '2009': SAMPLE_1_DIR / 'DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv',
    '2010': SAMPLE_1_DIR / 'DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv',
}

bene_year_ids = {
    year: set(pd.read_csv(path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])
    for year, path in bene_files.items()
}

bene_year_compare = pd.DataFrame([
    {
        'year': year,
        'beneficiaries': len(ids),
        'also_in_2008': len(ids & bene_year_ids['2008']),
        'not_in_2008': len(ids - bene_year_ids['2008']),
        'dropped_since_2008': len(bene_year_ids['2008'] - ids),
    }
    for year, ids in bene_year_ids.items()
])
bene_year_compare

,year,beneficiaries,also_in_2008,not_in_2008,dropped_since_2008
0,2008,116352,116352,0,0
1,2009,114538,114538,0,1814
2,2010,112754,112754,0,3598


### 2. Carrier Claims files (1A vs 1B)

CMS splits carrier claims into two files (A and B) because of file size. Same schema; **different claims** — no overlapping `CLM_ID` between A and B.

In [105]:
carrier_a = pd.read_csv(
    SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv',
    usecols=['DESYNPUF_ID', 'CLM_ID'],
    dtype=str,
)
carrier_b = pd.read_csv(
    SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv',
    usecols=['DESYNPUF_ID', 'CLM_ID'],
    dtype=str,
)

carrier_compare = pd.DataFrame([
    {
        'file': 'Carrier 1A',
        'rows': len(carrier_a),
        'unique_claims': carrier_a['CLM_ID'].nunique(),
        'unique_beneficiaries': carrier_a['DESYNPUF_ID'].nunique(),
    },
    {
        'file': 'Carrier 1B',
        'rows': len(carrier_b),
        'unique_claims': carrier_b['CLM_ID'].nunique(),
        'unique_beneficiaries': carrier_b['DESYNPUF_ID'].nunique(),
    },
])

carrier_compare['shared_CLM_ID_with_other_file'] = [0, 0]
carrier_compare

,file,rows,unique_claims,unique_beneficiaries,shared_CLM_ID_with_other_file
0,Carrier 1A,2370667,2370667,49278,0
1,Carrier 1B,2370668,2370668,49349,0


In [106]:
shared_carrier_clm = set(carrier_a['CLM_ID']) & set(carrier_b['CLM_ID'])
print(f'Shared CLM_ID between Carrier 1A and 1B: {len(shared_carrier_clm)}')
print('Same column schema:', list(pd.read_csv(SAMPLE_1_DIR / FILE_GROUPS["Carrier Claims (Part B)"][0], nrows=0).columns)
      == list(pd.read_csv(SAMPLE_1_DIR / FILE_GROUPS["Carrier Claims (Part B)"][1], nrows=0).columns))

Shared CLM_ID between Carrier 1A and 1B: 0
Same column schema: True


### 3. Inpatient vs Outpatient claims

Both are institutional claims with `CLM_ID`, but different settings, column counts, and claim volumes. No shared `CLM_ID` between inpatient and outpatient files.

In [107]:
ip_cols = set(pd.read_csv(SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv', nrows=0).columns)
op_cols = set(pd.read_csv(SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv', nrows=0).columns)

institutional_compare = pd.DataFrame([
    {'file': 'Inpatient', 'columns': len(ip_cols), 'shared_columns_with_other': len(ip_cols & op_cols)},
    {'file': 'Outpatient', 'columns': len(op_cols), 'shared_columns_with_other': len(ip_cols & op_cols)},
])
institutional_compare

,file,columns,shared_columns_with_other
0,Inpatient,81,74
1,Outpatient,76,74


In [108]:
print('Columns only in Inpatient:', sorted(ip_cols - op_cols))
print('Columns only in Outpatient:', sorted(op_cols - ip_cols))

Columns only in Inpatient: ['CLM_ADMSN_DT', 'CLM_DRG_CD', 'CLM_PASS_THRU_PER_DIEM_AMT', 'CLM_UTLZTN_DAY_CNT', 'NCH_BENE_DSCHRG_DT', 'NCH_BENE_IP_DDCTBL_AMT', 'NCH_BENE_PTA_COINSRNC_LBLTY_AM']
Columns only in Outpatient: ['NCH_BENE_PTB_COINSRNC_AMT', 'NCH_BENE_PTB_DDCTBL_AMT']


### 4. Prescription Drug Events (PDE)

Smallest column set (8 columns). One row per prescription fill (`PDE_ID`), not per claim. Covers Part D drug utilization from 2008–2010.

In [109]:
pde_cols = list(pd.read_csv(
    SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv',
    nrows=0,
).columns)

pd.DataFrame({'column': pde_cols, 'description': [
    'Beneficiary ID',
    'Prescription drug event ID (row key)',
    'Service / fill date',
    'Product/service ID (NDC-related)',
    'Quantity dispensed',
    'Days supply',
    'Patient pay amount',
    'Total prescription cost',
]})

,column,description
0,DESYNPUF_ID,Beneficiary ID
1,PDE_ID,Prescription drug event ID (row key)
2,SRVC_DT,Service / fill date
3,PROD_SRVC_ID,Product/service ID (NDC-related)
4,QTY_DSPNSD_NUM,Quantity dispensed
5,DAYS_SUPLY_NUM,Days supply
6,PTNT_PAY_AMT,Patient pay amount
7,TOT_RX_CST_AMT,Total prescription cost


### Are `PROD_SRVC_ID` values repeated per patient?

Each PDE row is one fill. For a given patient, we compare **total fills** vs **unique `PROD_SRVC_ID`** values. If fills > unique products, the same drug code appears on more than one row (a refill).

In [ ]:
PATIENT_ID = "DESYNPUF_ID"
PDE_PATH = SAMPLE_1_DIR / "DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv"

pde = pd.read_csv(
    PDE_PATH,
    usecols=[PATIENT_ID, "PDE_ID", "PROD_SRVC_ID", "SRVC_DT"],
    dtype={PATIENT_ID: str, "PDE_ID": str, "PROD_SRVC_ID": str},
    low_memory=False,
)

patient_products = pde.groupby(PATIENT_ID, observed=True).agg(
    n_fills=(PATIENT_ID, "count"),
    n_unique_products=("PROD_SRVC_ID", "nunique"),
)
patient_products["n_repeat_fills"] = patient_products["n_fills"] - patient_products["n_unique_products"]
patient_products["has_repeat_product"] = patient_products["n_fills"] > patient_products["n_unique_products"]

n_patients = len(patient_products)
n_with_repeats = patient_products["has_repeat_product"].sum()

print(f"PDE fills: {len(pde):,}")
print(f"Patients with PDE: {n_patients:,}")
print(f"Unique PROD_SRVC_ID overall: {pde['PROD_SRVC_ID'].nunique():,}")
print()
print(f"Patients with at least one repeated PROD_SRVC_ID: {n_with_repeats:,} ({100 * n_with_repeats / n_patients:.1f}%)")
print(f"Patients where every fill is a different product: {(~patient_products['has_repeat_product']).sum():,} ({100 * (1 - n_with_repeats / n_patients):.1f}%)")
print()
print("Per-patient summary (fills vs unique products):")
display(patient_products.describe().round(1))

# Among repeated patient-product pairs: different fill dates vs same day
repeat_pairs = (
    pde.groupby([PATIENT_ID, "PROD_SRVC_ID"], observed=True)
    .agg(n_fills=("SRVC_DT", "count"), n_unique_dates=("SRVC_DT", "nunique"))
    .reset_index()
)
repeat_pairs = repeat_pairs[repeat_pairs["n_fills"] > 1]
print(f"\nRepeated patient-product pairs: {len(repeat_pairs):,}")
print(f"  Different fill dates: {(repeat_pairs['n_unique_dates'] > 1).sum():,} ({100 * (repeat_pairs['n_unique_dates'] > 1).mean():.1f}%)")
print(f"  Same fill date only: {(repeat_pairs['n_unique_dates'] == 1).sum():,}")

# Example: patient with the most repeat fills
if n_with_repeats:
    example_id = patient_products.sort_values("n_repeat_fills", ascending=False).index[0]
    example = pde[pde[PATIENT_ID] == example_id].sort_values("SRVC_DT")
    print(f"\nExample patient {example_id}: {len(example)} fills, {example['PROD_SRVC_ID'].nunique()} unique products")
    display(example["PROD_SRVC_ID"].value_counts().head(5).to_frame("fill_count"))

### Map `PROD_SRVC_ID` to FDA NDC and check drug-level repeats

[`PROD_SRVC_ID`](https://open.fda.gov/apis/drug/ndc/how-to-use-the-endpoint/) is an 11-digit NDC-style code. We:

1. Convert it to **product NDC** (`#####-####`, first 9 digits) and **package NDC** (`#####-####-##`)
2. Query the [openFDA Drug NDC API](https://open.fda.gov/apis/drug/ndc/download/) in batches (cached locally)
3. Compare refill repeats at **package**, **product NDC**, and **FDA generic name** levels

**Note:** DE-SynPUF uses synthetic 2008-era codes; many may not appear in today's FDA directory. Product-level grouping still works without a match.

In [ ]:
import json
import ssl
import time
import urllib.error
import urllib.parse
import urllib.request

FDA_NDC_API = "https://api.fda.gov/drug/ndc.json"
FDA_CACHE_PATH = SAMPLE_1_DIR / "fda_ndc_lookup_cache.json"
FDA_BATCH_SIZE = 10           # product_ndc terms per API call
FDA_TOP_PRODUCTS_TO_QUERY = 200  # most common product NDC keys in Sample 1 PDE

_ssl_ctx = ssl.create_default_context()
_ssl_ctx.check_hostname = False
_ssl_ctx.verify_mode = ssl.CERT_NONE


def ndc11_to_product(ndc11: str) -> str:
    s = str(ndc11).zfill(11)
    return f"{s[:5]}-{s[5:9]}"


def ndc11_to_package(ndc11: str) -> str:
    s = str(ndc11).zfill(11)
    return f"{s[:5]}-{s[5:9]}-{s[9:11]}"


def load_fda_cache() -> dict:
    if FDA_CACHE_PATH.exists():
        return json.loads(FDA_CACHE_PATH.read_text())
    return {}


def save_fda_cache(cache: dict) -> None:
    FDA_CACHE_PATH.write_text(json.dumps(cache, indent=2))


def query_fda_product_ndcs(product_ndcs: list[str]) -> list[dict]:
    """Batch OR-search on product_ndc (openFDA requires quoted terms)."""
    if not product_ndcs:
        return []
    search = " OR ".join(f'product_ndc:"{ndc}"' for ndc in product_ndcs)
    params = urllib.parse.urlencode({
        "search": search,
        "limit": min(100, max(10, len(product_ndcs) * 5)),
    })
    url = f"{FDA_NDC_API}?{params}"
    req = urllib.request.Request(url, headers={"User-Agent": "medical-adherence-eda"})
    with urllib.request.urlopen(req, timeout=60, context=_ssl_ctx) as resp:
        payload = json.load(resp)
    if "error" in payload:
        return []
    return payload.get("results", [])


def lookup_fda_products(product_ndcs: list[str], cache: dict) -> dict:
    """Return {product_ndc: {generic_name, brand_name, package_ndcs}} using cache + API."""
    pending = [p for p in product_ndcs if p not in cache]
    for i in range(0, len(pending), FDA_BATCH_SIZE):
        batch = pending[i : i + FDA_BATCH_SIZE]
        try:
            results = query_fda_product_ndcs(batch)
        except (urllib.error.URLError, TimeoutError, json.JSONDecodeError) as exc:
            print(f"FDA API batch failed ({batch[0]} ...): {exc}")
            continue
        for rec in results:
            prod = rec.get("product_ndc")
            if not prod:
                continue
            packages = [p.get("package_ndc") for p in rec.get("packaging", []) if p.get("package_ndc")]
            cache[prod] = {
                "generic_name": rec.get("generic_name"),
                "brand_name": rec.get("brand_name"),
                "package_ndcs": packages,
            }
        for prod in batch:
            cache.setdefault(prod, None)
        time.sleep(0.3)
    return {prod: cache.get(prod) for prod in product_ndcs}


# Quick API sanity check (known real NDC in today's FDA directory)
_api_test = query_fda_product_ndcs(["65862-008"])
print("openFDA API check (65862-008 / metformin):", "OK" if _api_test else "no result")

# --- Attach NDC keys to PDE rows ---
pde_ndc = pde[[PATIENT_ID, "PROD_SRVC_ID"]].copy()
pde_ndc["product_ndc"] = pde_ndc["PROD_SRVC_ID"].map(ndc11_to_product)
pde_ndc["package_ndc"] = pde_ndc["PROD_SRVC_ID"].map(ndc11_to_package)

product_freq = pde_ndc["product_ndc"].value_counts()
top_products = product_freq.head(FDA_TOP_PRODUCTS_TO_QUERY).index.tolist()

fda_cache = load_fda_cache()
fda_lookup = lookup_fda_products(top_products, fda_cache)
save_fda_cache(fda_cache)

matched_products = [p for p, v in fda_lookup.items() if v]
print(f"\nQueried top {len(top_products):,} product NDC keys from Sample 1 PDE")
print(f"FDA matches: {len(matched_products):,} ({100 * len(matched_products) / len(top_products):.1f}% of queried)")

if matched_products:
    display(pd.DataFrame([
        {"product_ndc": p, **fda_lookup[p]} for p in matched_products[:10]
    ]))
else:
    print("No FDA matches for Sample 1 product codes (expected for DE-SynPUF synthetic NDCs).")

pde_ndc["generic_name"] = pde_ndc["product_ndc"].map(
    lambda p: (fda_lookup.get(p) or {}).get("generic_name") if fda_lookup.get(p) else pd.NA
)


def repeat_patient_stats(frame: pd.DataFrame, drug_col: str) -> pd.Series:
    fills = frame.groupby(PATIENT_ID, observed=True).size()
    uniq = frame.groupby(PATIENT_ID, observed=True)[drug_col].nunique()
    return pd.Series({
        "patients": fills.size,
        "mean_fills": round(fills.mean(), 1),
        "mean_unique_drugs": round(uniq.mean(), 1),
        "patients_with_repeat": int((fills > uniq).sum()),
        "pct_with_repeat": round(100 * (fills > uniq).mean(), 1),
    })


rows = []
for label, col in [
    ("Package NDC (PROD_SRVC_ID)", "PROD_SRVC_ID"),
    ("Product NDC (9-digit key)", "product_ndc"),
    ("FDA generic name (matched rows only)", "generic_name"),
]:
    sub = pde_ndc if col != "generic_name" else pde_ndc[pde_ndc["generic_name"].notna()]
    if sub.empty:
        rows.append({"level": label, "patients": 0, "pct_with_repeat": pd.NA, "note": "no FDA matches"})
    else:
        rows.append({"level": label, **repeat_patient_stats(sub, col)})

display(pd.DataFrame(rows))

print("Package = exact same PROD_SRVC_ID refilled")
print("Product = same 9-digit NDC product (different package sizes count as repeats)")
print("Generic = same FDA generic name (subset with API matches)")
print(f"Cache saved to: {FDA_CACHE_PATH}")

### 5. Side-by-side: what each file type contains

In [110]:
file_type_summary = pd.DataFrame([
    {
        'file_type': 'Beneficiary Summary',
        'what_it_records': 'Demographics, enrollment, chronic conditions, annual payment totals',
        'row_grain': '1 row per beneficiary per year',
        'row_key': 'DESYNPUF_ID',
        'time_span': 'Single calendar year (2008, 2009, or 2010)',
        'typical_use': 'Cohort definition, covariates, mortality',
    },
    {
        'file_type': 'Carrier Claims (1A + 1B)',
        'what_it_records': 'Physician/supplier services (Part B), diagnoses, procedures',
        'row_grain': '1 row per claim line',
        'row_key': 'CLM_ID (+ line fields)',
        'time_span': '2008–2010 (combined file)',
        'typical_use': 'Office visits, labs, physician diagnoses',
    },
    {
        'file_type': 'Inpatient Claims',
        'what_it_records': 'Hospital stays',
        'row_grain': '1 row per claim',
        'row_key': 'CLM_ID',
        'time_span': '2008–2010',
        'typical_use': 'Admissions, DRG, inpatient cost',
    },
    {
        'file_type': 'Outpatient Claims',
        'what_it_records': 'Hospital outpatient services',
        'row_grain': '1 row per claim',
        'row_key': 'CLM_ID',
        'time_span': '2008–2010',
        'typical_use': 'ER, outpatient procedures',
    },
    {
        'file_type': 'Prescription Drug Events',
        'what_it_records': 'Part D prescription fills',
        'row_grain': '1 row per fill',
        'row_key': 'PDE_ID',
        'time_span': '2008–2010',
        'typical_use': 'Medication adherence, drug utilization',
    },
])
file_type_summary

,file_type,what_it_records,row_grain,row_key,time_span,typical_use
0,Beneficiary Summary,"Demographics, enrollment, chronic conditions, ...",1 row per beneficiary per year,DESYNPUF_ID,"Single calendar year (2008, 2009, or 2010)","Cohort definition, covariates, mortality"
1,Carrier Claims (1A + 1B),"Physician/supplier services (Part B), diagnose...",1 row per claim line,CLM_ID (+ line fields),2008–2010 (combined file),"Office visits, labs, physician diagnoses"
2,Inpatient Claims,Hospital stays,1 row per claim,CLM_ID,2008–2010,"Admissions, DRG, inpatient cost"
3,Outpatient Claims,Hospital outpatient services,1 row per claim,CLM_ID,2008–2010,"ER, outpatient procedures"
4,Prescription Drug Events,Part D prescription fills,1 row per fill,PDE_ID,2008–2010,"Medication adherence, drug utilization"


### Summary: how Sample 1 files differ

| Difference | Detail |
|------------|--------|
| **Purpose** | Beneficiary = who they are; Claims = medical care received; PDE = drugs filled |
| **Row grain** | Beneficiary files: 1 row/person/year. Claims: many rows/person. PDE: many fills/person |
| **Columns** | 32 (bene) → 142 (carrier) → 81 (IP) / 76 (OP) → 8 (PDE) |
| **Time** | Bene files are annual snapshots; claims/PDE span 2008–2010 in one file |
| **Carrier 1A vs 1B** | Same layout, split for size — combine for full carrier claims |
| **2008 vs 2009 vs 2010 bene** | Same schema; fewer beneficiaries in later years (deaths/dropouts) |
| **Linking** | Join all files on `DESYNPUF_ID` within Sample 1 only |

## Are the same patients tracked over 2008–2010?

Compare beneficiary counts year over year: who stays in the cohort, who drops out, and whether anyone new is added.

In [111]:
bene_by_year = {
    2008: pd.read_csv(
        SAMPLE_1_DIR / 'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv',
        usecols=['DESYNPUF_ID', 'BENE_DEATH_DT'],
        dtype={'DESYNPUF_ID': str},
    ),
    2009: pd.read_csv(
        SAMPLE_1_DIR / 'DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv',
        usecols=['DESYNPUF_ID', 'BENE_DEATH_DT'],
        dtype={'DESYNPUF_ID': str},
    ),
    2010: pd.read_csv(
        SAMPLE_1_DIR / 'DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv',
        usecols=['DESYNPUF_ID', 'BENE_DEATH_DT'],
        dtype={'DESYNPUF_ID': str},
    ),
}

ids_by_year = {year: set(df['DESYNPUF_ID']) for year, df in bene_by_year.items()}

cohort_sizes = pd.Series(
    {year: len(ids) for year, ids in ids_by_year.items()},
    name='beneficiaries',
)
cohort_sizes

2008    116352
2009    114538
2010    112754
Name: beneficiaries, dtype: int64

In [112]:
in_all_three = ids_by_year[2008] & ids_by_year[2009] & ids_by_year[2010]

transitions = pd.DataFrame([
    {
        'period': '2008 → 2009',
        'starting_cohort': len(ids_by_year[2008]),
        'ending_cohort': len(ids_by_year[2009]),
        'stayed': len(ids_by_year[2008] & ids_by_year[2009]),
        'dropped': len(ids_by_year[2008] - ids_by_year[2009]),
        'added_new': len(ids_by_year[2009] - ids_by_year[2008]),
    },
    {
        'period': '2009 → 2010',
        'starting_cohort': len(ids_by_year[2009]),
        'ending_cohort': len(ids_by_year[2010]),
        'stayed': len(ids_by_year[2009] & ids_by_year[2010]),
        'dropped': len(ids_by_year[2009] - ids_by_year[2010]),
        'added_new': len(ids_by_year[2010] - ids_by_year[2009]),
    },
    {
        'period': '2008 → 2010',
        'starting_cohort': len(ids_by_year[2008]),
        'ending_cohort': len(ids_by_year[2010]),
        'stayed': len(ids_by_year[2008] & ids_by_year[2010]),
        'dropped': len(ids_by_year[2008] - ids_by_year[2010]),
        'added_new': len(ids_by_year[2010] - ids_by_year[2008]),
    },
])

transitions

,period,starting_cohort,ending_cohort,stayed,dropped,added_new
0,2008 → 2009,116352,114538,114538,1814,0
1,2009 → 2010,114538,112754,112754,1784,0
2,2008 → 2010,116352,112754,112754,3598,0


In [113]:
print(f'Beneficiaries present in all three years: {len(in_all_three):,}')
print(f'2009 is a subset of 2008: {ids_by_year[2009] <= ids_by_year[2008]}')
print(f'2010 is a subset of 2009: {ids_by_year[2010] <= ids_by_year[2009]}')
print(f'Any new patients added after 2008: {len(ids_by_year[2009] - ids_by_year[2008]) + len(ids_by_year[2010] - ids_by_year[2009]) > 0}')

Beneficiaries present in all three years: 112,754
2009 is a subset of 2008: True
2010 is a subset of 2009: True
Any new patients added after 2008: False


### Why do patients drop out?

Check whether dropped beneficiaries have a death date recorded in the prior year's file.

In [114]:
dropped_08_to_09 = ids_by_year[2008] - ids_by_year[2009]
dropped_09_to_10 = ids_by_year[2009] - ids_by_year[2010]

death_checks = pd.DataFrame([
    {
        'drop_period': '2008 → 2009',
        'dropped_count': len(dropped_08_to_09),
        'with_death_date_in_prior_year_file': bene_by_year[2008]
            .loc[bene_by_year[2008]['DESYNPUF_ID'].isin(dropped_08_to_09), 'BENE_DEATH_DT']
            .notna()
            .sum(),
    },
    {
        'drop_period': '2009 → 2010',
        'dropped_count': len(dropped_09_to_10),
        'with_death_date_in_prior_year_file': bene_by_year[2009]
            .loc[bene_by_year[2009]['DESYNPUF_ID'].isin(dropped_09_to_10), 'BENE_DEATH_DT']
            .notna()
            .sum(),
    },
])

death_checks

,drop_period,dropped_count,with_death_date_in_prior_year_file
0,2008 → 2009,1814,1814
1,2009 → 2010,1784,1784


In [115]:
# Visualize cohort flow
cohort_flow = pd.DataFrame({
    'group': [
        'In all 3 years',
        'Dropped by 2009 only',
        'Dropped by 2010 (was in 2009)',
    ],
    'count': [
        len(in_all_three),
        len(dropped_08_to_09),
        len(dropped_09_to_10),
    ],
})
cohort_flow

,group,count
0,In all 3 years,112754
1,Dropped by 2009 only,1814
2,Dropped by 2010 (was in 2009),1784


### Summary

| Question | Answer |
|----------|--------|
| Same patients tracked? | **Yes** — no new `DESYNPUF_ID` values are added after 2008 |
| Cohort pattern | Each later year is a **subset** of the prior year (monotone attrition) |
| In all 3 years | **112,754** beneficiaries |
| Dropped 2008 → 2009 | **1,814** (all had `BENE_DEATH_DT` in the 2008 file) |
| Dropped 2009 → 2010 | **1,784** (all had `BENE_DEATH_DT` in the 2009 file) |
| Total dropped 2008 → 2010 | **3,598** (≈3.1% of the 2008 cohort) |
| Added | **0** new patients in 2009 or 2010 |

Dropouts align with recorded deaths — the DE-SynPUF cohort is a **closed panel** that shrinks over time, not an open cohort with new enrollments.

## Do IDs match across sample folders and claim files?

Checks:
1. **Sample folders** (Sample 1 vs Sample 2, …) — same ID pools?
2. **Sample 1 files** — same `DESYNPUF_ID` set in every file?
3. **Claim files** — does patient ID map across Carrier 1A, Carrier 1B, Inpatient, Outpatient, and PDE?

In [116]:
CLAIM_FILES = {
    'Carrier 1A': SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Carrier_Claims_Sample_1A.csv',
    'Carrier 1B': SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Carrier_Claims_Sample_1B.csv',
    'Inpatient': SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv',
    'Outpatient': SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv',
    'PDE': SAMPLE_1_DIR / 'DE1_0_2008_to_2010_Prescription_Drug_Events_Sample_1.csv',
}

BENE_2008 = SAMPLE_1_DIR / 'DE1_0_2008_Beneficiary_Summary_File_Sample_1 2.csv'

claim_ids = {
    name: set(pd.read_csv(path, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])
    for name, path in CLAIM_FILES.items()
}
bene_2008_ids = set(pd.read_csv(BENE_2008, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])

pd.Series({name: len(ids) for name, ids in claim_ids.items()}, name='unique_DESYNPUF_ID')

Carrier 1A    49278
Carrier 1B    49349
Inpatient     37780
Outpatient    85272
PDE           99538
Name: unique_DESYNPUF_ID, dtype: int64

### 1. IDs across sample folders (Sample 1, Sample 2, …)

In [117]:
sample_folder_ids = {}
for folder in sorted(DATA_ROOT.glob('Sample *'), key=lambda p: int(p.name.split()[-1])):
    bene_file = next(folder.glob('*2008*Beneficiary*'))
    sample_folder_ids[folder.name] = set(
        pd.read_csv(bene_file, usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID']
    )

sample_1_folder_ids = sample_folder_ids['Sample 1']
overlaps_with_other_samples = {
    folder: len(sample_1_folder_ids & ids)
    for folder, ids in sample_folder_ids.items()
    if folder != 'Sample 1'
}

pd.Series(overlaps_with_other_samples, name='shared_ids_with_sample_1')

Sample 2     0
Sample 3     0
Sample 4     0
Sample 5     0
Sample 6     0
Sample 7     0
Sample 8     0
Sample 9     0
Sample 10    0
Sample 11    0
Sample 12    0
Sample 13    0
Sample 14    0
Sample 15    0
Sample 16    0
Sample 17    0
Sample 18    0
Sample 19    0
Sample 20    0
Name: shared_ids_with_sample_1, dtype: int64

In [118]:
print('IDs identical across all sample folders:', all(v == 0 for v in overlaps_with_other_samples.values()))
print('Sample 1 uses a separate ID pool from every other sample folder.')

IDs identical across all sample folders: True
Sample 1 uses a separate ID pool from every other sample folder.


### 2. IDs across all Sample 1 files

Every claim/PDE ID must exist in the 2008 Beneficiary Summary. Files do **not** all contain the same ID set (different beneficiaries appear in different claim types).

In [119]:
id_match_checks = []

for name, ids in claim_ids.items():
    orphans = ids - bene_2008_ids
    id_match_checks.append({
        'file': name,
        'unique_ids': len(ids),
        'ids_in_2008_bene': len(ids & bene_2008_ids),
        'orphans_not_in_bene': len(orphans),
        'maps_to_bene_summary': len(orphans) == 0,
        'exact_same_set_as_2008_bene': ids == bene_2008_ids,
    })

id_match_df = pd.DataFrame(id_match_checks)
id_match_df

,file,unique_ids,ids_in_2008_bene,orphans_not_in_bene,maps_to_bene_summary,exact_same_set_as_2008_bene
0,Carrier 1A,49278,49278,0,True,False
1,Carrier 1B,49349,49349,0,True,False
2,Inpatient,37780,37780,0,True,False
3,Outpatient,85272,85272,0,True,False
4,PDE,99538,99538,0,True,False


In [120]:
union_claim_ids = set().union(*claim_ids.values())

print(f'2008 Beneficiary Summary IDs: {len(bene_2008_ids):,}')
print(f'Union of IDs across all 5 claim/PDE files: {len(union_claim_ids):,}')
print(f'Same set (bene == union of claims): {bene_2008_ids == union_claim_ids}')

2008 Beneficiary Summary IDs: 116,352
Union of IDs across all 5 claim/PDE files: 103,935
Same set (bene == union of claims): False


### 3. Does `DESYNPUF_ID` map across Carrier 1A, 1B, Inpatient, Outpatient, and PDE?

**Yes for joins** — same ID format, all valid in the beneficiary file.

**No for identical coverage** — each file contains a different subset of beneficiaries. Carrier 1A and 1B split patients almost entirely between files (not duplicate patients).

In [121]:
file_names = list(claim_ids.keys())
overlap_matrix = pd.DataFrame(index=file_names, columns=file_names, dtype=int)

for a in file_names:
    for b in file_names:
        overlap_matrix.loc[a, b] = len(claim_ids[a] & claim_ids[b])

overlap_matrix

,Carrier 1A,Carrier 1B,Inpatient,Outpatient,PDE
Carrier 1A,49278.0,1.0,18904.0,42330.0,47127.0
Carrier 1B,1.0,49349.0,18820.0,42381.0,47134.0
Inpatient,18904.0,18820.0,37780.0,36314.0,36480.0
Outpatient,42330.0,42381.0,36314.0,85272.0,81492.0
PDE,47127.0,47134.0,36480.0,81492.0,99538.0


In [122]:
in_all_claim_files = set.intersection(*claim_ids.values())
in_carrier_a_and_b = claim_ids['Carrier 1A'] & claim_ids['Carrier 1B']

carrier_split = pd.DataFrame([
    {'group': 'In all 5 claim/PDE files', 'beneficiaries': len(in_all_claim_files)},
    {'group': 'In both Carrier 1A and 1B', 'beneficiaries': len(in_carrier_a_and_b)},
    {'group': 'Only in Carrier 1A', 'beneficiaries': len(claim_ids['Carrier 1A'] - claim_ids['Carrier 1B'])},
    {'group': 'Only in Carrier 1B', 'beneficiaries': len(claim_ids['Carrier 1B'] - claim_ids['Carrier 1A'])},
])
carrier_split

,group,beneficiaries
0,In all 5 claim/PDE files,1
1,In both Carrier 1A and 1B,1
2,Only in Carrier 1A,49277
3,Only in Carrier 1B,49348


In [123]:
# Example: one beneficiary with records in multiple file types
multi_file_benes = claim_ids['Outpatient'] & claim_ids['PDE'] & claim_ids['Inpatient']
example_id = next(iter(multi_file_benes))

print(f'Example DESYNPUF_ID: {example_id}')
for name, ids in claim_ids.items():
    print(f'  {name}: {"yes" if example_id in ids else "no"}')
print(f'  2008 Beneficiary Summary: {"yes" if example_id in bene_2008_ids else "no"}')

Example DESYNPUF_ID: D50117A56318BBDC
  Carrier 1A: no
  Carrier 1B: yes
  Inpatient: yes
  Outpatient: yes
  PDE: yes
  2008 Beneficiary Summary: yes


### Summary

| Question | Answer |
|----------|--------|
| Same IDs in all **sample folders**? | **No** — Sample 1–20 are separate cohorts with **zero** shared `DESYNPUF_ID` values |
| Same IDs in every **Sample 1 file**? | **No** — each file has a different count; claims files are subsets of the 116,352 beneficiaries |
| Do all claim IDs exist in Beneficiary Summary? | **Yes** — 0 orphan IDs in Carrier 1A/1B, Inpatient, Outpatient, PDE |
| Does patient ID **map** across claim files? | **Yes** — use `DESYNPUF_ID` to join; same ID always refers to the same person within Sample 1 |
| Carrier 1A vs 1B | **Different patients** in each file (~49k each, only **1** in both); combine both files for full carrier coverage |
| Inpatient vs Outpatient vs PDE | Overlapping but **not identical** patient sets; many patients appear in only some file types |

## Why can PDE unique IDs exceed Carrier 1A + 1B?

Carrier files are **Part B physician/supplier claims** only. PDE is **Part D prescription fills**. These are different Medicare programs — a beneficiary can have drug fills without a carrier (office visit) claim in the same 2008–2010 window.

In [124]:
carrier_a_ids = set(pd.read_csv(CLAIM_FILES['Carrier 1A'], usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])
carrier_b_ids = set(pd.read_csv(CLAIM_FILES['Carrier 1B'], usecols=['DESYNPUF_ID'], dtype=str)['DESYNPUF_ID'])
pde_ids = claim_ids['PDE']
ip_ids = claim_ids['Inpatient']
op_ids = claim_ids['Outpatient']

carrier_union = carrier_a_ids | carrier_b_ids

carrier_pde_compare = pd.DataFrame([
    {'metric': 'Unique IDs in Carrier 1A', 'count': len(carrier_a_ids)},
    {'metric': 'Unique IDs in Carrier 1B', 'count': len(carrier_b_ids)},
    {'metric': 'In both Carrier 1A and 1B', 'count': len(carrier_a_ids & carrier_b_ids)},
    {'metric': 'Carrier 1A ∪ 1B (union, not sum)', 'count': len(carrier_union)},
    {'metric': 'Unique IDs in PDE', 'count': len(pde_ids)},
    {'metric': 'PDE IDs not in carrier union', 'count': len(pde_ids - carrier_union)},
    {'metric': 'Carrier union IDs not in PDE', 'count': len(carrier_union - pde_ids)},
])
carrier_pde_compare

,metric,count
0,Unique IDs in Carrier 1A,49278
1,Unique IDs in Carrier 1B,49349
2,In both Carrier 1A and 1B,1
3,"Carrier 1A ∪ 1B (union, not sum)",98626
4,Unique IDs in PDE,99538
5,PDE IDs not in carrier union,5278
6,Carrier union IDs not in PDE,4366


In [125]:
# Naive sum double-counts the 1 beneficiary in both carrier files
naive_sum = len(carrier_a_ids) + len(carrier_b_ids)
print(f'Naive sum (1A + 1B): {naive_sum:,}  ← overcounts by {naive_sum - len(carrier_union):,}')
print(f'Correct union (1A ∪ 1B): {len(carrier_union):,}')
print(f'PDE exceeds carrier union by: {len(pde_ids) - len(carrier_union):,}')

Naive sum (1A + 1B): 98,627  ← overcounts by 1
Correct union (1A ∪ 1B): 98,626
PDE exceeds carrier union by: 912


In [126]:
pde_without_carrier = pde_ids - carrier_union
any_medical = carrier_union | ip_ids | op_ids

pde_gap_breakdown = pd.DataFrame([
    {
        'group': 'PDE but no Carrier 1A/1B',
        'beneficiaries': len(pde_without_carrier),
    },
    {
        'group': '… but has Outpatient claim',
        'beneficiaries': len(pde_without_carrier & op_ids),
    },
    {
        'group': '… but has Inpatient claim',
        'beneficiaries': len(pde_without_carrier & ip_ids),
    },
    {
        'group': '… has Outpatient or Inpatient (no carrier)',
        'beneficiaries': len(pde_without_carrier & (op_ids | ip_ids)),
    },
    {
        'group': 'PDE with no Carrier, Inpatient, or Outpatient',
        'beneficiaries': len(pde_ids - any_medical),
    },
])
pde_gap_breakdown

,group,beneficiaries
0,PDE but no Carrier 1A/1B,5278
1,… but has Outpatient claim,533
2,… but has Inpatient claim,54
3,… has Outpatient or Inpatient (no carrier),569
4,"PDE with no Carrier, Inpatient, or Outpatient",4709


### Why this is expected (not a data error)

Your intuition — *"you need a visit to get a prescription"* — applies to **new** prescriptions, but Medicare data does not require a **Part B carrier claim** for every **Part D fill**:

1. **Different claim types** — Carrier = Part B (physician/office). PDE = Part D (pharmacy). They are billed and stored separately.
2. **Refills & renewals** — Ongoing medications can be refilled at a pharmacy without a new office visit in 2008–2010.
3. **Other care settings** — Some patients have **outpatient or inpatient** claims but no carrier claim; drugs may still appear in PDE.
4. **PDE-only beneficiaries** — ~4,700 Sample 1 beneficiaries have Part D fills but **no** carrier, inpatient, or outpatient claim in this dataset (still valid in CMS synthetic data; not everyone with drugs has a matching physician claim in the extract).
5. **Carrier 1A + 1B is not "all medical care"** — it is only ~85% of beneficiaries with any PDE activity, by design of who appears in the carrier split files.

**For joins:** use `DESYNPUF_ID` across files — IDs still map correctly. **For analysis:** do not assume PDE ⊆ Carrier; use left joins and expect PDE-only and carrier-only patients.